pkg/handlers/authorization/roles_service.go

```diff
@@ -0,0 +1,66 @@
+package authorization
+
+import (
+	"time"
+
+	"github.com/bluele/gcache"
+	"github.com/openshift/elasticsearch-proxy/pkg/clients"
+	"github.com/openshift/elasticsearch-proxy/pkg/config"
+	"github.com/openshift/elasticsearch-proxy/pkg/handlers"
+	log "github.com/sirupsen/logrus"
+)
+
+type rolesService struct {
+	cache gcache.Cache
+}
+
+func NewRolesService(size int, expiry time.Duration, roleConfig map[string]config.BackendRoleConfig, client clients.OpenShiftClient) *rolesService {
+	return &rolesService{
+		cache: gcache.New(size).
+			LRU().
+			Expiration(expiry).
+			LoaderFunc(fetchRolesFn(roleConfig, client)).
+			Build(),
+	}
+}
+
+type reviewRoles struct {
+	review *clients.TokenReview
+	roles  map[string]bool
+}
+
+func (s *rolesService) getRolesAndReviews(token string) (*reviewRoles, error) {
+	v, err := s.cache.Get(token)
+	if err != nil {
+		return nil, err
+	}
+	cacheVal := v.(*reviewRoles)
+	return cacheVal, nil
+}
+
+func fetchRolesFn(roleConfig map[string]config.BackendRoleConfig, client clients.OpenShiftClient) func(key interface{}) (interface{}, error) {
+	return func(key interface{}) (interface{}, error) {
+		token := key.(string)
+		tokenReview, err := client.TokenReview(token)
+		if err != nil {
+			log.Errorf("Error fetching user info %v", err)
+			return nil, err
+		}
+		ctx := &handlers.RequestContext{}
+		ctx.UserName = tokenReview.UserName()
+		log.Debugf("User is %q", tokenReview.UserName())
+
+		roles := map[string]bool{}
+		for name, sar := range roleConfig {
+			if allowed, err := client.SubjectAccessReview(tokenReview.UserName(), sar.Namespace, sar.Verb, sar.Resource, sar.ResourceAPIGroup); err == nil {
+				log.Debugf("%q for %q SAR: %v", tokenReview.UserName(), name, allowed)
+				roles[name] = true
```

Comment:
Could this not be a set?  Is there a need for us to keep the rolename to a value that is always true?  Can we not assume that if a role is in the set then it is allowed?

Reply:
What data structure would you like to use here? 

A set seems ideal as later we want to access by name. We can change it to `map[bool]struct{}` which should be the best for a set.

Reply:
> What data structure would you like to use here?
> 
> A set seems ideal as later we want to access by name. We can change it to `map[bool]struct{}` which should be the best for a set.

We can use `"k8s.io/apimachinery/pkg/util/sets"` unless we dont have that vendored otherwise `map[string]bool` is fine

Reply:
I have changed to `map[bool]struct{}` it's the preferred way to define a set in go. Rather use core language than another import.